module/zfs/mmp.c

```diff
@@ -0,0 +1,548 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or http://www.opensolaris.org/os/licensing.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+/*
+ * Copyright (c) 2017 by Lawrence Livermore National Security, LLC.
+ */
+
+#include <sys/abd.h>
+#include <sys/dsl_pool.h>
+#include <sys/mmp.h>
+#include <sys/spa.h>
+#include <sys/spa_impl.h>
+#include <sys/vdev.h>
+#include <sys/vdev_impl.h>
+#include <sys/zfs_context.h>
+#include <sys/callb.h>
+
+/*
+ * MMP (Multi-Modifier Protection) attempts to prevent a user from
+ * importing or opening a pool on more than one host at a time, in
+ * particular by issuing "zpool import -f" on one host, while the pool is
+ * already imported on another host.  There are many other ways in which
+ * a device could be used by two hosts for different purposes at the same
+ * time, resulting in pool damage.  This implementation does not attempt
+ * to detect most of those cases.
+ * 
+ * MMP operates by ensuring there is frequent visible change on disk (a
+ * "heartbeat") at all times, and altering the import process to check
+ * for such change and failing the import if it is detected.
+ * 
+ * On-disk:
+ * 
```

Comment:
nit: You're editor is leaving whitespace at the end of these empty comment lines.  Please fix.